### ISSUE TITLE
Help tracking down rare case of corruption on save

### ISSUE THREAD

**Issue by innerthought** [musescore/MuseScore#17057]
### Issue type

File corruption

### Bug description

Saved files become corrupted on opening. "Open anyway" doesn't work. I want to stress that all the files were saved and closed during the previous session without any message like "File cannot be saved properly" or "Can't save the file"

### Steps to reproduce

1. got to: either of the files attached
2. then: open it and try to click "Open anyway"
3. then: see that nothing happens
[cuba_01.mscz.zip](https://github.com/musescore/MuseScore/files/11097350/cuba_01.mscz.zip)
[progression_14_of_13.mscz.zip](https://github.com/musescore/MuseScore/files/11097353/progression_14_of_13.mscz.zip)
[Added on: 2023.04.03] [cuba_11_of_10.mscz.zip](https://github.com/musescore/MuseScore/files/11133636/cuba_11_of_10.mscz.zip)

With each file, while trying to open it, I firstly get this message:

"This file contains errors that could cause MuseScore to malfunction"

and when I click on "Show details", I get this: 

Error=XML_ERROR_EMPTY_DOCUMENT ErrorID=13 (0xd) Line number=0

### MuseScore Version

4.0.2.230651553

### Regression

Yes, this used to work in a previous version of Musescore 4.x

### Operating system

OS: Win-10, 64-bit operating system, x64-based processor

**Comment by MarcSabatella** [comment]
Both of these files open without incident for me.  They also continue to open normally if I saved and reopen.  Are you ure you attached the correct versions of these files?

**Comment by innerthought** [comment]
Yes, I attached the correct versions.
Strange. I just downloaded them from this page and now I can open them. Why is it so? Is it because I added ".zip" extension while uploading and then, after downloading, got rid of ".zip"?

**Comment by MarcSabatella** [comment]
That shouldn’t make any difference to the content of the file, but if your drive starting to fail, that sort of random error could be a sign of it.

**Comment by cbjeukendrup** [comment]
Here's a theory: in the folder that originally contained the files, there might be .autosave versions of these files, maybe because MuseScore failed to remove them or because it crashed. When such .autosave files are present and you try to open the file, MuseScore will always open the autosave file instead. Maybe the autosave files were corrupted, so therefore the files can't be opened. When you downloaded the files from this page, they were most likely placed in your Downloads folder, which does not contain the autosave files of course, so therefore MuseScore tries to open the files themselves, which succeeds. 

Is that possible?

**Comment by innerthought** [comment]
@cbjeukendrup "Here's a theory: in the folder that originally ... Is that possible?"
- Yes, that's exactly what happened.

**Comment by innerthought** [comment]
This is just amazing! It turns out that I've stumbled upon the way how to recover a file from simply uploading it here (which is impossible without adding ".zip" extension to the file name)!  :)
 
Now I do the same to all my crashed files that I can't open by MuseScore4 software: 
1. I firstly rename it by adding ".zip" to the name. For example, if the corrupted file name is "kitten.mscz" then it will be "kitten.mscz.zip" now;
2. Then, by using 7-Zip software I use "Extract to" option - thus, creating "kitten.mscz" folder;
3. Then, I look into "kitten.mscz" folder and find there the file named as "kitten.mscx". The extension ".mscx" is different from my usual MuseScore files (which is  ".mscz"), but the Type is still specified as "MuseScore File", so it can be opened with MuseScore;
4. So, I open it with MuseScore and I see my file with all the modifications performed during the last session!;
5. Then I simply re-save it as "kitten.mscz" in the original folder. The file is recovered and saved!

**Comment by MarcSabatella** [comment]
I think that is @cbjeukendrup is right about what's happening, it's much simpler than that - just delete the corrupt autosave.  Or rename your scores.

What you suggest here won't *really* work.  The MSCX file doesn't contain all of the score data - it's missing style and playback settings, etc.  So, you'll still be losing work,

**Comment by innerthought** [comment]
@MarcSabatella Hmm, strange. I thought I was able to recover all the latest work. Well, I may be wrong  because it's been a while since the latest work was done and I might've forgotten. 

Okay, I'll try the way you are suggesting.

So, should I now close this issue? (I am not familiar with proper way of doing things on GitHub)

**Comment by innerthought** [comment]
@MarcSabatella: Marc, are you sure what you are saying is correct? The thing is that my experience is directly opposite. Last night I modified and saved one of my files and opened it this morning - only to see what has already become a usual thing: all the panning and volumes gone and some instruments switched back to MS Basic in the mixer. Then I tried my newly discovered way: added ".zip" to the file, extracted with 7-Zip, opened the extracted folder and opened the .mscx file in it. The file contained all the modifications I performed last night, including panning, volumes and all the instrument specifications.  So all I had to do now was just to save this file as an .mscz file, which I did and, thus, restored all my last night's work.

**Comment by innerthought** [comment]
@cbjeukendrup "MuseScore will always open the autosave file instead"
-- On second thought, if it were autosave files that MuseScore tied to open first then I would have gotten something like "File "cuba_03_of_02.mscz.autosave" is corrupted", but what I get is this:
![corrupted](https://user-images.githubusercontent.com/119835219/229115289-991b4f26-3b76-47d2-b7fe-6b5a56264ce1.jpg)
There is no ".autosave" extension, which means it it not an autosave file that MuseScore is struggling with.
Besides, "Open anyway" still produces no effect, which means it's not about which folder I am currently in, but probably the software's problem.

**Comment by innerthought** [comment]
@MarcSabatella:

> “It could be a direct result of having previously manipulated that MSCZ file in an unsupported way”

-- All I did was adding some measures before and after selection and then adding some notes to the new empty measures. I remember, though, that I was adding some note by pasting. Could pasting be an unsupported way?

> “or it could be another symptom of a failing drive”

-- Perhaps, yes. But then the question is why I have this symptom only when I run MS4 and not when I run any other app. I didn’t have this even with MS3.

> “It could also be an issue with your Muse Sounds installation if the score was using those and suddenly doesn't seem to be anymore.”

-- I have just deleted MS4 and all the MuseSounds and then reinstalled both again – the problem still persists.

> “Best to ask for help with that on the Support forum” 

-- Strange, I thought GitHub was the place with technical experts and developers, while the Support forum, as I thought, was mostly a place with long-time users. Okay, I'll do, of course.

> “I don't think that proves anything. Since you didn't specifically ask for the autosave file - MuseScore was apparently substituting the autosave internally - it makes sense that error message would report the originally-specified filename.”

> -- Do you mean that the message may be saying ‘File “name.mscz” is corrupted’ while, in fact, it’s a “name.mscz.autosave” file?

Can you try and see if you can open my “cuba_11_of_10.mscz”, which I have added to the first message above just now? I saved it yesterday and today I can’t open it even using the way that I have described earlier.

**Comment by innerthought** [comment]
@cbjeukendrup:

> “MuseScore says the filename without .autosave extension, but in reality it is really trying to open the autosave file.”

-- But from time to time I do see the message saying ‘File “name.mscz.autosave” is corrupted’ as opposed to the ‘File “name.mscz” is corrupted’ message.

> “is it possible that this was caused by [MU4 Issue] When file is "dirty" because of changes in the Mixer, AutoSave marks it "saved" #16549?”

-- I don’t know. Maybe. The problem described there is very similar to mine, except in my case I often see the asterisk even right after openning the file, that is, not even modifying it or playing it back. In other words, I see the asterisk even though I haven’t done anything to the file besides accessing it.

**Comment by MarcSabatella** [comment]
> -- All I did was adding some measures before and after selection and then adding some notes to the new empty measures. I remember, though, that I was adding some notes by pasting. Could pasting be an unsupported way?

No, I mean directly manipulating the MSCZ file via a third party ZIP program.  If you tried extracting files from MSCZ archives and opening them directly, or manually placing a file into the MSCZ archive, this could mess things up in unpredictable ways.

> > “Best to ask for help with that on the Support forum”
> 
> -- Strange, I thought GitHub was the place with technical experts and developers, while the Support forum, as I thought, was mostly a place with long-time users

It’s true that GitHub is where developers are, but their job is to, well, develop :-).  So the only thing that belongs here are confirmed bug reports with clear steps to reproduce.

There isn’t a paid support staff; support is provided on via the Support forum.  There is a ton of technical expertise there even if it doesn’t typically come from the paid development staff.  We volunteers can usually help figure out steps to reproduce a problem and pinpoint whether it is a bug or not, and then if so, a clear bug report can be posted here.

> -- Do you mean that the message may be saying ‘File “name.mscz” is corrupted’ while, in fact, it’s a “name.mscz.autosave” file?

Precisely.  There might well be some cases where MuseScore displays the autosave name and other cases where it suppressed, depending on how exactly it was opened.  So no conclusions can be drawn from that.

> Can you try and see if you can open my “cuba_11_of_10.mscz”, which I added to the first message above just now?

Yes, it is completely normal as far as I can see.

**Comment by innerthought** [comment]
@MarcSabatella:

> Yes, it is completely normal as far as I can see. 

- You still can open it?! Unbelievable! Can you, please, resend it to me here? At least, I'll have a copy that I know is still openable, except on other systems.

**Comment by innerthought** [comment]
@MarcSabatella:

- I just installed MuseSounds and MuseScore on a laptop and donwloaded this file. Tried to open it - to no avail. The file is still corrupted and "Open anyway" doesn't work on that laptop in this case either. Well, at least, from this I know that this is not my desktop's system problem.

I have also [reported this issue on MuseScore support forum](https://musescore.org/en/node/347667), and people there can't open this file either.

**Comment by MarcSabatella** [comment]
I've responded there.  It's not the same file as I downloaded from here - I think I got the wrong one since there were two versions there - but indeed, that one is corrupt.

I think though your post title and description is inaccurate, and that's going to lead to confusion.  The problem doesn't happen on *open*, it happens on save, and obviously, not for all files, just under some very specific circumstances that so far you havebn't been able to figure out.  So what you're look for here is, help figuring out the steps to reproduce that rare corruption, so that you can then come and file a proper issue report here.  So a better title might be "help tracking down rare case of corruption on save", and the description should make it clear this issue is very unique to something about your system and/or your process and that's you're looking for helping tracking down what makes some percentage of your saves work differently from all others.

**Comment by innerthought** [comment]
@MarcSabatella 

> I think though your post title and description is inaccurate, and that's going to lead to confusion
-- Thank you. I changed the title to yours just now. 
The reason why I thought the file was being corrupted on opening and not on saving was because I never had a "file cannot be saved properly" message and, on the other hand, the file often opened with an asterisk prepended to its name - as if I had already performed some changes while all I did was just opening it.

**Comment by MarcSabatella** [comment]
Seeing an asterisk implies the score *did* in fact open, so any cases you saw of that would be entirely unrelated to this issue.  But most likely, that's just a matter of the score having been imported (e.g., from a previous version) and the "change" being noted s your election whether to update style settings etc.  I could also imagine it happening if the original sounds used aren't present and so MuseScore is needing to substitute others.  Again, entirely unrelated, but if you're looking for helping tracking that down, a separate forum post devoted to that would be the way to go.  The goal being, to either ascertain that it is something simple like that, or to identify a specific score that for some reason triggers this despite nothing having changed.  So that if there is a bug and someone can confirm, then you can open a proper issue here for that as well.

**Comment by innerthought** [comment]
@MarcSabatella:

> Seeing an asterisk implies the score did in fact open, so any cases you saw of that would be entirely unrelated to this issue

-- I see. Thank you for the explanation.

**Comment by MarcSabatella** [comment]
To be clear, it's on the forum where you need the better title/description, because that's where you're asking for help.

**Comment by innerthought** [comment]
@MarcSabatella:  

> ...it's on the forum where you need the better title/description

-- Ah, okay. I'll change the title there, too.

**Comment by cbjeukendrup** [comment]
I'm closing this, since it's a very lengthy discussion that is not workable for the developers, and there has not been any activity for a while.

**Comment by innerthought** [comment]
> I'm closing this, since it's a very lengthy discussion that is not workable for the developers...

- So, you've closed this thread just because it was lenghthy and "not workable" while the issue is still not solved?!

> ...and there has not been any activity for a while.

- I wonder what kind of further activity do expect from us users? All we can do is just report the issue and humbly waiting when the issue in question is tackled by the developers. In fact, it is us users waiting for some activity. You, however, made the fact that no actitivity followed as valid reason for not solving the issue.

**Comment by cbjeukendrup** [comment]
> > I'm closing this, since it's a very lengthy discussion that is not workable for the developers...
> 
> -- So, you've closed this thread just because it was lenghthy and "not workable" while the issue is still not solved?!

Well, that is what I mean: the thread is so lengthy and unclear that it isn't even clear if there is still any issue!

If there is still any issue, please discuss it on MuseScore.org, and if you and others come to the conclusion that there is indeed a concrete, reproducible bug in MuseScore, then feel free to create a new issue, that is clear and concise enough to be looked at by developers.

**Comment by innerthought** [comment]
> If there is still any issue, please discuss it on MuseScore.org
-- The issue has never been fixed and it was already discussed on MuseScore.org around the time this issue was posted: https://musescore.org/en/node/347667

Just re-opened this issue:
https://github.com/musescore/MuseScore/issues/18087

**Comment by MarcSabatella** [comment]
The issue has not been fixed because you haven't yet given the steps to reproduce the issue.  That's what we have been trying to explain from the beginning - you should only open an issue on GitHub once you have precise steps to reproduce the problem.  Until then, *please* don't use GitHub.  If you need help trying to figure out how to reproduce a problem, use the Support forum.  Yes, you started a discussion there, but it seems you still haven't found the steps.  So continue the discussion there and only there until you are able to find steps to reproduce the problem.

Only once you are able to find the precise steps to reproduce a problem should you be opening GitHub issues.  Not following these guidelines just wastes time and makes it harder for the developers to focus on actually fixing the issues that *are* posted with precise steps to reproduce.
